ir/base/smart.go

```diff
@@ -0,0 +1,93 @@
+package base
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/libp2p/go-smart-record/ir"
+)
+
+//NOTE: Including everything in the same file for discussion purposes.
+// This should be organized in its corresponding files when implementing it.
+type SmartNode interface {
+	ir.Node
+	Disassemble() ir.Node // returns only syntactic nodes
+}
+
+// SmartString is the smart node for the String type.
+// The difference is that additional metadata is included in the assembly process
+type SmartString struct {
+	Value    ir.String
+	Metadata metadataContext
+}
+
+// List of metadata attributes supported.
+type metadataContext struct {
+	ttl uint64
+}
+
+// TTL sets in metadata
+func TTL(value uint64) Metadata {
+	return func(m *metadataContext) error {
+		m.ttl = value
+		return nil
+	}
+}
+
+// Option type for smart records
+type Metadata func(*metadataContext) error
+
+// Applies metadata items to a metadataContext
+func (m *metadataContext) apply(items ...Metadata) error {
+	for i, item := range items {
+		if err := item(m); err != nil {
+			return fmt.Errorf("error applying metadata value: %s", i, err)
+		}
+	}
+	return nil
+}
+
+// The Assembler includes metadata to the syntactic node and assembles a smartNode.
+// SmartNode includes additional metadata
+// This belongs to (asm SequenceAssembler)
+func Assemble(ctx ir.AssemblerContext, src ir.Node, metadata ...Metadata) (SmartNode, error) {
```

Comment:
I think we should go with this interface. Some comments:

This interface implies that the user applies one metadata to all nodes in given update.
I think this is the common case and natural pattern for devs.

In the (distant) future, there may or may not arise a need to specify metadata per-node in a single update. If that happens we can add a second mechanism for specifying TTLs that is e.g. based on asking the user to include an optional "metadata" field in the (syntactic) dictionary nodes of their update. Just FYI. I doubt this will be necessary in the next 3 years.